crates/flux-desugar/src/desugar.rs

```diff
@@ -1581,21 +1633,65 @@ fn index_sort(
     resolver_output: &ResolverOutput,
     bty: &surface::BaseTy,
 ) -> Option<fhir::Sort> {
-    // CODESYNC(sort-of, 4) sorts should be given consistently
+    // CODESYNC(sort-of, 3) sorts should be given consistently
     match &bty.kind {
-        surface::BaseTyKind::Path(path) => {
-            let res = resolver_output.path_res_map[&path.node_id];
-            genv.sort_of_res(res)
-        }
+        surface::BaseTyKind::Path(path) => sort_of_surface_path(genv, resolver_output, path),
         surface::BaseTyKind::Slice(_) => Some(fhir::Sort::Int),
     }
 }
 
-fn as_tuple<'a>(genv: &'a GlobalEnv, sort: &'a fhir::Sort) -> &'a [fhir::Sort] {
-    if let fhir::Sort::Record(def_id) = sort {
-        genv.index_sorts_of(*def_id)
+fn sort_of_surface_path(
+    genv: &GlobalEnv,
+    resolver_output: &ResolverOutput,
+    path: &surface::Path,
+) -> Option<fhir::Sort> {
+    // CODESYNC(sort-of-path, 2) sorts should be given consistently
+    let res = resolver_output.path_res_map[&path.node_id];
+
+    match res {
+        fhir::Res::PrimTy(PrimTy::Int(_) | PrimTy::Uint(_)) => Some(fhir::Sort::Int),
+        fhir::Res::PrimTy(PrimTy::Bool) => Some(fhir::Sort::Bool),
+        fhir::Res::PrimTy(PrimTy::Float(..) | PrimTy::Str | PrimTy::Char) => Some(fhir::Sort::Unit),
+        fhir::Res::Def(DefKind::TyAlias { .. } | DefKind::Enum | DefKind::Struct, def_id) => {
+            // TODO: duplication with sort_of_path
+            let mut sort_args = vec![];
+            if let Some(generics) = genv.map().get_generics(def_id) {
+                for (param, arg) in generics.params.iter().zip(&path.generics) {
+                    if let GenericParamKind::SplTy = param.kind {
+                        let surface::GenericArg::Type(ty) = arg else { return None };
+                        let surface::BaseTyKind::Path(path) = &ty.as_bty()?.kind else {
+                            return None;
+                        };
+                        let sort = sort_of_surface_path(genv, resolver_output, path)?;
+                        sort_args.push(sort);
+                    }
+                }
+            };
+            Some(fhir::Sort::Record(def_id, List::from_vec(sort_args)))
+        }
+        fhir::Res::Def(DefKind::TyParam, def_id) => {
+            let param = genv.get_generic_param(def_id.expect_local());
+            match &param.kind {
+                fhir::GenericParamKind::BaseTy | fhir::GenericParamKind::SplTy => {
```

Comment:
`fhir::Sort::Param` represents the sort of a parameter of kind `BaseTy`. `SplTy` does not have a sort in this sense. I tested this and all test pass.

It should be possible to have a concept of a sort associated to a `SplTy` but that should be a different variant. Note that returning `None` for `SplTy` is conservative as we would be rejecting things that could potentially be accepted.


```suggestion
                fhir::GenericParamKind::BaseTy => {
```

Reply:
done!